### DISCUSSION TITLE
Consider combining multiple mockito related modules as one for integration testing

### DISCUSSION THREAD

**Discussion by ZSmallX** [robolectric/robolectric#7109] [Ideas]
After this PR:  

https://github.com/robolectric/robolectric/pull/7086  

There are 3 modules related with mockito: `mockito`, `mockito-kotlin` and `mockito-experimental`. It brings several potential problems:  
- ambiguous naming because some features are not experimental in the latest mockito.
- separeted code and redundant runtime configuration. Though runtime classpath isolation can be good in some situations.

Once we combine those modules as one, we can:  
- cleaner arch to find mockito related tests in one module, just like `powermock`, `mockk` etc.
- remove ambiguous and stale naming.
- share one runtime classpath and configuration.

**Comment by utzcoz** [comment]
`mockito-experimental` is used to test `mocikito-inline` firstly, and it contains some normal mockito related tests. And `mockito` is used to test `mockito-core` with related tests. `mockito-kotlin` is used to test mockito with Kotlin functions in Robolectric environment with `mockito-core`. Maybe we can combine `mockito` and `mockito-kotlin` with name `mockito`, and use this new module to test mokito-core related  compatibility with Java and Kotlin. We also can move normal mockito tests from `mockito-experimental` to new `mockito` module and leave `mockito-inline` related tests to `mockito-experimental`, and rename `mockito-experimental` to `mockito-inline`. cc /@hoisie

**Reply by hoisie** [reply]
Does it make sense to just have one mockito project that only uses mockito-inline? So we would move all of the tests to one project and that would use mockito-inline. From my understanding mockito inline and non-inline use the same instrumentation (bytebuddy-based), it's just that the inline one installs a java agent and hooks into the Instrumentation API. So in some sense mockito-inline tests already cover all of the functionality of the non-inline case.

**Reply by utzcoz** [reply]
Looks great to me. I don't know what the exact different between `mockito-core` and `mockito-inline` before. If `mockito-inline` can cover what `mockito-core` does, we can combine them to one module with `mockito-inline` to reduce tech debt. Could you help to combine those modules to one @ZSmallX ?

**Reply by ZSmallX** [reply]
yeah, I would work on it, but I got little into `mockito-core` and `mockito-inlne` before. And I will do more investigations on it and combine those modules as one once they were confirmed to work the same for Robolectric integration tests.

**Reply by hoisie** [reply]
On second thought maybe it would just be simpler to keep two separate mockito integration tests. One called `mockito`, and one called `mockito-inline`.
